source/ref_gl/r_renderer.h

```diff
@@ -0,0 +1,173 @@
+#include "Extensions/NRIMeshShader.h"
+#include "Extensions/NRIRayTracing.h"
+#include "Extensions/NRISwapChain.h"
+
+#define DECLARE_RENDERER_FUNCTION( ret, name, ... ) \
+	typedef ret( qcdecl *name##Fn )( __VA_ARGS__ ); \
+	extern name##Fn name;
+
+#define GAMMARAMP_STRIDE 4096
+
+typedef enum { 
+	BACKEND_OPENGL_LEGACY, 
+	BACKEND_NRI_VULKAN, 
+	BACKEND_NRI_METAL, 
+	BACKEND_NRI_DX12 
+} backend_api_t;
+
+typedef struct {
+	int _extMarker;
+
+	//
+	// only uint8_ts must follow the extensionsBoolMarker
+	//
+	char draw_range_elements;
+	char multitexture;
+	char texture_cube_map;
+	char texture_edge_clamp;
+	char texture_filter_anisotropic;
+	char texture_compression;
+	char compressed_ETC1_RGB8_texture;
+	char vertex_buffer_object;
+	char GLSL;
+	char GLSL_core;
+	char GLSL130;
+	char depth_texture;
+	char framebuffer_object;
+	char vertex_shader;
+	char fragment_shader;
+	char shader_objects;
+	char shading_language_100;
+	char shading_language_130;
+	char bgra;
+	char gamma_control;
+	char swap_control;
+	char draw_instanced;
+	char instanced_arrays;
+	char gpu_memory_info;
+	char meminfo;
+	char framebuffer_blit;
+	char depth24;
+	char depth_nonlinear;
+	char get_program_binary;
+	char rgb8_rgba8;
+	char ES3_compatibility;
+	char blend_func_separate;
+	char texture_array;
+	char fragment_precision_high;
+	char packed_depth_stencil;
+	char texture_lod;
+	char gpu_shader5;
+
+	union {
+		char shadow, shadow_samplers;
+	};
+	union {
+		char texture3D, texture_3D;
+	};
+	union {
+		char texture_non_power_of_two, texture_npot;
+	};
+	union {
+		char half_float_vertex, vertex_half_float;
+	};
+} glextinfo_t;
+
+typedef struct {
+	const char *rendererString;
+	const char *vendorString;
+	const char *versionString;
+	const char *extensionsString;
+	const char *glwExtensionsString;
+	const char *shadingLanguageVersionString;
+	unsigned versionHash;
+
+	const char *applicationName;
+	const char *screenshotPrefix;
+	int startupColor;
+
+	int version;
+	int shadingLanguageVersion;
+
+	int width, height;
+	bool fullScreen;
+
+	bool stereoEnabled;
+	int stencilBits;
+
+	bool hwGamma;
+	unsigned short gammaRampSize;
+	unsigned short originalGammaRamp[3 * GAMMARAMP_STRIDE];
+
+	float depthEpsilon;
+
+	int maxTextureSize;
+	int maxTextureUnits;
+	int maxTextureCubemapSize;
+	int maxTexture3DSize;
+	int maxTextureLayers;
+	int maxTextureFilterAnisotropic;
+	int maxRenderbufferSize;
+	int maxVaryingFloats;
+	int maxVertexUniformComponents;
+	int maxVertexAttribs;
+	int maxFragmentUniformComponents;
+	unsigned int maxGLSLBones; // the maximum amount of bones we can handle in a vertex shader
+
+	bool forceRGBAFramebuffers; // PowerVR hack - its blending interprets alpha in RGB FBs as 0, not 1
+	bool multithreading;
+
+	glextinfo_t ext;
+} glconfig_t;
+
+typedef struct {
+	backend_api_t backend;
+	union {
+		struct {
+			NriDevice *device;
+		} nri;
+		struct {
+			glconfig_t config;
+		} gl;
+	};
+} renderer_t;
+
+extern renderer_t renderer;
```

Comment:
umm I don't quite like how this is global umm maybe I shouldn't need to pass around render_t everywhere.